File: rails/app/models/article.rb
class Article < ApplicationRecord
  self.table_name = 'articles'

  enum :status, { draft: 0, published: 1, archived: 2 }, prefix: :status
  enum :article_type, { guide: 0, tierlist: 1, matchup: 2, news: 3, spotlight: 4, decklist: 5 }, prefix: :article_type
  enum :language, { e_n: 0, d_e: 1, f_r: 2, i_t: 3, e_s: 4, j_p: 5, p_t: 6 }, prefix: :language

  has_many :tag_assignments, class_name: 'ArticleTagAssignment', inverse_of: :article
  has_many :comments, class_name: 'ArticleComment', inverse_of: :article
  belongs_to :author, class_name: 'Player', inverse_of: :articles
  belongs_to :featured_deck, class_name: 'Deck', inverse_of: :articles, optional: true
  has_many :tags, class_name: 'ArticleTag', through: :tag_assignments

  validates :title, presence: true, length: { maximum: 300 }
  validates :slug, presence: true, length: { maximum: 300 }
  validates :slug, uniqueness: { message: 'slug must be unique' }

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:view_count_not_negative, 'Article view count must not be negative') unless ((view_count.nil? || view_count >= 0))
    errors.add(:likes_count_not_negative, 'Article likes count must not be negative') unless ((likes_count.nil? || likes_count >= 0))
  end

  # Domain invariants — IMPLIES rules
  validate :validate_implies

  def validate_implies
    errors.add(:base, 'Published article must have a published_at timestamp') if (status == 'published') && published_at.nil?
  end

  def to_s
    title.to_s
  end

  # Business operations

  def publish
    # TODO: implement publish
  end

  def archive
    # TODO: implement archive
  end

  def increment_view
    # TODO: implement increment_view
  end

  def like
    # TODO: implement like
  end

  def unlike
    # TODO: implement unlike
  end

  def reading_time_minutes
    # TODO: implement reading_time_minutes
    nil
  end

  # Lifecycle state machine
  ALLOWED_TRANSITIONS = {
    'draft' => ['published'],
    'published' => ['archived'],
    'archived' => ['draft'],
  }.freeze

  def assert_transition!(to_state)
    allowed = ALLOWED_TRANSITIONS.fetch(status.to_s, [])
    unless allowed.include?(to_state.to_s)
      raise ArgumentError, "Transition #{status} -> #{to_state} not allowed"
    end
  end

  # Lifecycle hooks
  after_save :update_search_index

  def update_search_index
    # TODO: implement update_search_index
  end

  def as_json(options = {})
    hash = super(options)
    hash['createdAt'] = hash.delete('created_at') if hash.key?('created_at')
    hash['updatedAt'] = hash.delete('updated_at') if hash.key?('updated_at')
    hash['publishedAt'] = hash.delete('published_at') if hash.key?('published_at')
    hash
  end
end

File: dotnet/CardsProject/Domain/Content/Article.cs
using CardsProject.Domain.Players;
using CardsProject.Domain.Cards;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace CardsProject.Domain.Content;

public enum ArticleStatusType
{
    Draft,
    Published,
    Archived
}

public enum ArticleArticleTypeType
{
    Guide,
    Tierlist,
    Matchup,
    News,
    Spotlight,
    Decklist
}

public enum ArticleLanguageType
{
    EN,
    DE,
    FR,
    IT,
    ES,
    JP,
    PT
}

public class Article : IValidatableObject
{
    public int Id { get; set; }

    public string Title { get; set; } = "";
    public string Slug { get; set; } = "";
    public string Body { get; set; } = "";
    public string? Excerpt { get; set; }
    public string? CoverImageUrl { get; set; }
    public ArticleStatusType Status { get; set; }
    public ArticleArticleTypeType ArticleType { get; set; }
    public ArticleLanguageType Language { get; set; }
    public int ViewCount { get; set; } = 0;
    public int LikesCount { get; set; } = 0;
    public bool IsFeatured { get; set; } = false;
    [JsonPropertyName("publishedAt")]
    public DateTime? PublishedAt { get; set; } = null;
    [JsonPropertyName("createdAt")]
    public DateTime? CreatedAt { get; set; } = null;
    [JsonPropertyName("updatedAt")]
    public DateTime? UpdatedAt { get; set; } = null;

    public int? AuthorId { get; set; }
    [ForeignKey(nameof(AuthorId))]
    public Player? Author { get; set; }
    public int? FeaturedDeckId { get; set; }
    [ForeignKey(nameof(FeaturedDeckId))]
    public Deck? FeaturedDeck { get; set; }

    public ICollection<ArticleTagAssignment> TagAssignments { get; set; } = new List<ArticleTagAssignment>();
    public ICollection<ArticleComment> Comments { get; set; } = new List<ArticleComment>();

    // Business operations

    public void Publish()
    {
        // TODO: implement publish
    }

    public void Archive()
    {
        // TODO: implement archive
    }

    public void IncrementView()
    {
        // TODO: implement increment_view
    }

    public void Like()
    {
        // TODO: implement like
    }

    public void Unlike()
    {
        // TODO: implement unlike
    }

    public int ReadingTimeMinutes()
    {
        // TODO: implement reading_time_minutes
        return default;
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static readonly System.Collections.Generic.Dictionary<ArticleStatusType, ArticleStatusType[]> AllowedTransitions = new()
    {
        [ArticleStatusType.Draft] = new[] { ArticleStatusType.Published },
        [ArticleStatusType.Published] = new[] { ArticleStatusType.Archived },
        [ArticleStatusType.Archived] = new[] { ArticleStatusType.Draft }
    };

    public void AssertTransition(ArticleStatusType to)
    {
        if (!AllowedTransitions.TryGetValue(Status, out var allowed) || !System.Array.Exists(allowed, s => s == to))
            throw new InvalidOperationException($"Transition {Status} -> {to} not allowed");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( ViewCount >= 0 ))
            yield return new ValidationResult("Article view count must not be negative", new[] { nameof(Id) });
        if (!( LikesCount >= 0 ))
            yield return new ValidationResult("Article likes count must not be negative", new[] { nameof(Id) });
    }

    // ── Lifecycle hooks (call from AppDbContext.SaveChangesAsync override) ───
    public void UpdateSearchIndex()
    {
        // TODO: implement update_search_index
    }
}
